src/driver-holtek/holtek8-shared.c

```diff
@@ -0,0 +1,1291 @@
+
+	if (sensor_cfg->zero_indexed)
+		raw += 1;
+
+	return raw * sensor_cfg->dpi_step;
+}
+
+static bool
+holtek8_button_data_is_equal(const struct holtek8_button_data *lhs, const struct holtek8_button_data *rhs)
+{
+	if (lhs->type != rhs->type)
+		return false;
+
+	for (unsigned int i = 0; i < sizeof(lhs->data); ++i) {
+		if (lhs->data[i] != rhs->data[i]) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static bool
+holtek8_button_map_action_to_raw(const struct ratbag_button_action *action, struct holtek8_button_data *data)
+{
+	const struct holtek8_button_mapping *mapping = NULL;
+
+	ARRAY_FOR_EACH(holtek8_button_map, mapping) {
+		if (!ratbag_button_action_match(&mapping->action, action)) {
+			continue;
+		}
+
+		*data = mapping->data;
+		return true;
+	}
+
+	return false;
+}
+
+static const struct ratbag_button_action *
+holtek8_button_map_raw_to_action(const struct holtek8_button_data *data)
+{
+	const struct holtek8_button_mapping *mapping = NULL;
+
+	ARRAY_FOR_EACH(holtek8_button_map, mapping) {
+		if (!holtek8_button_data_is_equal(data, &mapping->data))
+			continue;
+
+		return &mapping->action;
+	}
+
+	return NULL;
+}
+
+struct holtek8_modifier_mapping {
+	enum holtek8_modifiers modifiers;
+	unsigned int key;
+};
+
+static const struct holtek8_modifier_mapping holtek8_modifier_map[] = {
+	{ HOLTEK8_MODIFIER_LEFTCTRL, KEY_LEFTCTRL },
+	{ HOLTEK8_MODIFIER_LEFTSHIFT, KEY_LEFTSHIFT },
+	{ HOLTEK8_MODIFIER_LEFTALT, KEY_LEFTALT },
+	{ HOLTEK8_MODIFIER_LEFTMETA, KEY_LEFTMETA },
+	{ HOLTEK8_MODIFIER_RIGHTCTRL, KEY_RIGHTCTRL },
+	{ HOLTEK8_MODIFIER_RIGHTSHIFT, KEY_RIGHTSHIFT },
+	{ HOLTEK8_MODIFIER_RIGHTALT, KEY_RIGHTALT },
+	{ HOLTEK8_MODIFIER_RIGHTMETA, KEY_RIGHTMETA },
+};
+
+/*
+ * Reads macro events from the device memory.
+ *
+ * Macros in memory are divided into pages. If the device supports jumping between
+ * pages, tries to follow jumps to read multipaged macros.
+ *
+ * @return 0 on success or a negative errno.
+ */
+static int
+holtek8_read_macro_data(struct ratbag_device *device, struct holtek8_macro_event *macro_events, uint8_t macro_idx)
+{
+	int rc;
+	struct holtek8_data *drv_data = device->drv_data;
+	struct holtek8_feature_report report = {0, 0, {macro_idx}, 0};
+	struct holtek8_macro_data macro_data;
+	unsigned int data_i = 0, events_i = 0;
+	bool single_page_macros = false;
+
+	switch (drv_data->api_version) {
+		case HOLTEK8_API_A:
+			report.command = HOLTEK8A_CMD_READ_MACRO_DATA;
+			single_page_macros = true;
+			break;
+		case HOLTEK8_API_B:
+			report.command = HOLTEK8B_CMD_READ_MACRO_DATA;
+			break;
+	}
+
+	rc = holtek8_set_feature_report(device, &report);
+	if (rc < 0)
+		return rc;
+
+	rc = holtek8_read_chunked(device, (uint8_t*) &macro_data, HOLTEK8_MACRO_DATA_SIZE, NULL);
+	if (rc < 0)
+		return rc;
+
+	while (events_i < HOLTEK8_MAX_MACRO_EVENTS) {
+		struct holtek8_macro_event *ev = &macro_data.event[data_i];
+		uint8_t *ev_data = (uint8_t*) ev;
```

Comment:
Like below:
https://github.com/libratbag/libratbag/pull/1561/files#r1468901974
Consider making holtek8_macro_event a union with uint8_t array or maybe even another struct>> (considering the `ev_data[1] == HOLTEK8_MACRO_CMD_JUMP *here*)